### PR TITLE
Cross Platform: ByteSwap

### PR DESCRIPTION
**Pull request by obastemur**
Define ByteSwapX for `__APPLE__` and `__linux__` targets

### CODE REVIEW

File: lib/common/common/ByteSwap.h [hunk 1]
```diff
@@ -50,6 +48,22 @@ RtlUlonglongByteSwap(
...
+#elif defined(__linux__)
+#include <byteswap.h>
+
+#define RtlUshortByteSwap(_x)    __bswap_16((_x))
+#define RtlUlongByteSwap(_x)     __bswap_32((_x))
+#define RtlUlonglongByteSwap(_x) __bswap_64((_x))
+
+#else
+// TODO: for BSD use endian.h?
+THIS PLATFORM IS NOT SUPPORTED
```
Comment (caitp) [inline lib/common/common/ByteSwap.h]: `#error "ByteSwapX not implemented on this platform"` maybe?

Comment (obastemur) [inline lib/common/common/ByteSwap.h]: Sure it looks better. Updated the PR accordingly.

File: lib/common/common/ByteSwap.h [hunk 2]
```diff
@@ -21,9 +21,7 @@ extern "C" {
 #define RtlUshortByteSwap(_x)    _byteswap_ushort((USHORT)(_x))
 #define RtlUlongByteSwap(_x)     _byteswap_ulong((_x))
 #define RtlUlonglongByteSwap(_x) _byteswap_uint64((_x))
-#else
-
-#ifdef _WIN32
+#elif defined (_WIN32)
```
Comment (jianchun) [inline lib/common/common/ByteSwap.h]: nit pick: Please remove the space in ```defined (_WIN32)```.

File: lib/common/common/ByteSwap.h [hunk 3]
```diff
@@ -50,6 +48,22 @@ RtlUlonglongByteSwap(
     _In_ ULONGLONG Source
     );
 #endif
-#endif
 
+#elif defined(__APPLE__)
```
Comment (jianchun) [inline lib/common/common/ByteSwap.h]: How is __APPLE__ get defined? ```https://github.com/dotnet/coreclr/blob/master/CMakeLists.txt``` deals with ```CLR_CMAKE_PLATFORM_DARWIN``` but not adding a symbol. Maybe we need to add one to ours?

Comment (caitp) [inline lib/common/common/ByteSwap.h]: it's predefined by GCC and Clang, and probably older Apple toolchains too (eg https://github.com/llvm-mirror/clang/blob/ea88b232d6f5178898909ecd39688b5c675d39ab/lib/Basic/Targets.cpp#L115)

File: lib/common/common/ByteSwap.h [hunk 4]
```diff
@@ -50,6 +48,22 @@ RtlUlonglongByteSwap(
...
+#include <libkern/OSByteOrder.h> 
+
+#define RtlUshortByteSwap(_x)    OSSwapInt16((_x))
+#define RtlUlongByteSwap(_x)     OSSwapInt32((_x))
+#define RtlUlonglongByteSwap(_x) OSSwapInt64((_x))
+
+#elif defined(__linux__)
```
Comment (jianchun) [inline lib/common/common/ByteSwap.h]: Similar question about ```__linux__```. https://github.com/dotnet/coreclr/blob/master/CMakeLists.txt does defining ```__LINUX__```. Maybe we need to add that to our CMakeLists.txt. Would ```__LINUX__``` also work for you?

File: lib/common/common/ByteSwap.h (hunk 3 above)
Comment (obastemur) [inline lib/common/common/ByteSwap.h]: > How is __APPLE__ get defined?
It's there by default on OSX/iOS..

File: lib/common/common/ByteSwap.h (hunk 4 above)
Comment (obastemur) [inline lib/common/common/ByteSwap.h]: `__linux__` is a predefined macro by GCC on Linux.

File: lib/common/common/ByteSwap.h (hunk 3 above)
Comment (jianchun) [inline lib/common/common/ByteSwap.h]: @caitp @obastemur Thanks for the info!

File: lib/common/common/ByteSwap.h (hunk 4 above)
Comment (jianchun) [inline lib/common/common/ByteSwap.h]: > __linux__ is a predefined macro by GCC on Linux.

Thanks! Seems also defined by clang. This is good then.

**Comment by obastemur** [comment]
@jianchun 
 > nit pick: Please remove the space in defined (_WIN32).

Done. and PR is updated

**Pull request merged**